velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,356 @@
+  // array(> kMediumSort).
+  // The two threshold values are given in the reference paper "Engineering a
+  // Sort Function".
+  static const int kSmallSort = 7;
+  static const int kMediumSort = 40;
+
+  template <typename TCompare>
+  void quickSort(char* start, char* end, TCompare compare) const {
+    quickSort(
+        detail::PrefixSortIterator(start, entrySize_),
+        detail::PrefixSortIterator(end, entrySize_),
+        compare);
+  }
+
+  /// For testing only.
+  template <typename TCompare>
+  FOLLY_ALWAYS_INLINE static char* testingMedian3(
+      char* a,
+      char* b,
+      char* c,
+      const uint64_t entrySize,
+      TCompare cmp) {
+    return *median3(
+        detail::PrefixSortIterator(a, entrySize),
+        detail::PrefixSortIterator(b, entrySize),
+        detail::PrefixSortIterator(c, entrySize),
+        cmp);
+  }
+
+ private:
+  FOLLY_ALWAYS_INLINE void swap(
+      const detail::PrefixSortIterator& lhs,
+      const detail::PrefixSortIterator& rhs) const {
+    simd::memcpy(swapBuffer_, *lhs, entrySize_);
+    simd::memcpy(*lhs, *rhs, entrySize_);
+    simd::memcpy(*rhs, swapBuffer_, entrySize_);
+  }
+
+  FOLLY_ALWAYS_INLINE void rangeSwap(
+      const detail::PrefixSortIterator& start1,
+      const detail::PrefixSortIterator& start2,
+      uint64_t length) const {
+    for (uint64_t i = 0; i < length; i++) {
+      swap(start1 + i, start2 + i);
+    }
+  }
+
+  /// Calculate which one has median of three input iterators.
+  /// The compare logic is (the symbol '<' and '>' means value compare result):
+  ///            ---------cmp(a, b)--------
+  ///           |                          |
+  ///        a<b|                      a>=b|
+  ///       cmp(b, c)                  cmp(b, c)
+  ///       |       |                  |       |
+  ///    b<c|   b>=c|               b>c|   b<=c|
+  ///   a<=b<=c  cmp(a, c)        c<=b<=a   cmp(a, c)
+  ///             |      |                   |      |
+  ///          a<c|  a>=c|                a>c|  a<=c|
+  ///         a<=c<=b  c<=a<=b           b<=c<=a  b<=a<=c
+  /// case a<=b=<c return b
+  /// case a<=c=<b return c
+  /// case c<=a<=b return a
+  /// case c<=b<=a return b
+  /// case b<=c<=a return c
+  /// case b<=a<=c return a.
+  template <typename TCompare>
+  FOLLY_ALWAYS_INLINE static detail::PrefixSortIterator median3(
+      const detail::PrefixSortIterator& a,
+      const detail::PrefixSortIterator& b,
+      const detail::PrefixSortIterator& c,
+      TCompare cmp) {
+    return cmp(*a, *b) < 0 ? (cmp(*b, *c) < 0       ? b
+                                  : cmp(*a, *c) < 0 ? c
+                                                    : a)
+                           : (cmp(*b, *c) > 0       ? b
+                                  : cmp(*a, *c) > 0 ? c
+                                                    : a);
+  }
+
+  template <typename TCompare>
+  FOLLY_ALWAYS_INLINE void insertSort(
```

Comment:
Do we consider using sorting network here?  Of course if we plan to use radix sort later this is not so important, but still can provide an better baseline.

Reply:
> Do we consider using sorting network here
Not so far, I hope to do algorithm level opt at last (after major part of normalize keys are implemented)